File: sdk/typescript/packages/api/src/queue.ts
export type OverflowPolicy = 'drop_oldest' | 'drop_newest';

export interface QueueOptions {
  maxSize?: number;
  batchSize?: number;
  flushIntervalMs?: number;
  overflowPolicy?: OverflowPolicy;
}

export type FlushCallback<T> = (items: T[]) => Promise<void>;

// EventQueue is a generic event buffer with configurable batching and overflow policies.
// Provider implementations use this to buffer events before sending them in batches.
export class EventQueue<T> {
  private readonly items: T[] = [];
  private readonly maxSize: number;
  private readonly batchSize: number;
  private readonly flushIntervalMs: number;
  private readonly overflowPolicy: OverflowPolicy;
  private readonly onFlush: FlushCallback<T>;
  private timer: ReturnType<typeof setInterval> | undefined;

  constructor(onFlush: FlushCallback<T>, opts: QueueOptions = {}) {
    this.onFlush = onFlush;
    this.maxSize = opts.maxSize ?? 10000;
    this.batchSize = opts.batchSize ?? 100;
    this.flushIntervalMs = opts.flushIntervalMs ?? 10000;
    this.overflowPolicy = opts.overflowPolicy ?? 'drop_oldest';

    if (this.flushIntervalMs > 0) {
      this.timer = setInterval(() => {
        void this.flushAll();
      }, this.flushIntervalMs);
    }
  }

  enqueue(item: T): void {
    if (this.items.length >= this.maxSize) {
      if (this.overflowPolicy === 'drop_oldest') {
        this.items.shift();
      } else {
        return;
      }
    }
    this.items.push(item);
    if (this.items.length >= this.batchSize) {
      void this.flush();
    }
  }

  async flush(): Promise<void> {
    if (this.items.length === 0) return;
    const batch = this.items.splice(0, this.batchSize);
    await this.onFlush(batch);
  }

  async flushAll(): Promise<void> {
    while (this.items.length > 0) {
      await this.flush();
    }
  }

  async shutdown(): Promise<void> {
    if (this.timer !== undefined) {
      clearInterval(this.timer);
      this.timer = undefined;
    }
    await this.flushAll();
  }

  get size(): number {
    return this.items.length;
  }
}

File: sdk/kotlin/api/src/main/kotlin/io/eventspec/analytics/Queue.kt
package io.eventspec.analytics

import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

enum class OverflowPolicy {
  DROP_OLDEST,
  DROP_NEWEST
}

data class QueueOptions(
    val maxSize: Int = 10_000,
    val batchSize: Int = 100,
    val flushIntervalMs: Long = 10_000,
    val overflowPolicy: OverflowPolicy = OverflowPolicy.DROP_OLDEST,
)

// EventQueue buffers events and flushes them in batches.
class EventQueue<T>(
    private val onFlush: suspend (List<T>) -> Unit,
    private val opts: QueueOptions = QueueOptions(),
    scope: CoroutineScope = CoroutineScope(Dispatchers.Default),
) {
  private val items = ArrayDeque<T>()
  private val mutex = Mutex()
  private var timer: Job? = null

  init {
    if (opts.flushIntervalMs > 0) {
      timer =
          scope.launch {
            while (isActive) {
              delay(opts.flushIntervalMs)
              flushAll()
            }
          }
    }
  }

  suspend fun enqueue(item: T) {
    mutex.withLock {
      if (items.size >= opts.maxSize) {
        when (opts.overflowPolicy) {
          OverflowPolicy.DROP_OLDEST -> items.removeFirst()
          OverflowPolicy.DROP_NEWEST -> return
        }
      }
      items.addLast(item)
    }
    if (size() >= opts.batchSize) flush()
  }

  suspend fun flush() {
    val batch =
        mutex.withLock {
          if (items.isEmpty()) return
          val b = items.take(opts.batchSize).toList()
          repeat(b.size) { items.removeFirst() }
          b
        }
    onFlush(batch)
  }

  suspend fun flushAll() {
    while (size() > 0) flush()
  }

  suspend fun shutdown() {
    timer?.cancel()
    flushAll()
  }

  fun size(): Int = items.size
}
